[Merged by Bors] - Cascaded shadow maps.

danchia opened the pull request:
Co-authored-by: Robert Swain <robert.swain@gmail.com>

# Objective

Implements cascaded shadow maps for directional lights, which produces better quality shadows without needing excessively large shadow maps.

Fixes #3629

Before
![image](https://user-images.githubusercontent.com/1222141/210061203-bbd965a4-8d11-4cec-9a88-67fc59d0819f.png)

After
![image](https://user-images.githubusercontent.com/1222141/210061334-2ff15334-e6d7-4a31-9314-f34a7805cac6.png)


## Solution

Rather than rendering a single shadow map for directional light, the view frustum is divided into a series of cascades, each of which gets its own shadow map. The correct cascade is then sampled for shadow determination.

---

## Changelog

Directional lights now use cascaded shadow maps for improved shadow quality.


## Migration Guide

You no longer have to manually specify a `shadow_projection` for a directional light, and these settings should be removed. If customization of how cascaded shadow maps work is desired, modify the `CascadeShadowConfig` component instead.

mockersf commented:
Nice!

Your branch is 78 commits late from main and has conflicts, could you resolve those?

danchia commented:
> Nice!
> 
> Your branch is 78 commits late from main and has conflicts, could you resolve those?

Rebased + fixed a bug to do with shader indexing.

superdump commented on examples/tools/scene_viewer/scene_viewer_plugin.rs:
```diff
@@ -198,35 +198,13 @@ fn keyboard_animation_control(
     mut query: Query<(&mut Transform, &mut DirectionalLight)>,
     mut animate_directional_light: Local<bool>,
 ) {
-    let mut projection_adjustment = Vec3::ONE;
-    if key_input.just_pressed(KeyCode::Key5) {
-        projection_adjustment.x -= SCALE_STEP;
-    } else if key_input.just_pressed(KeyCode::Key6) {
-        projection_adjustment.x += SCALE_STEP;
-    } else if key_input.just_pressed(KeyCode::Key7) {
-        projection_adjustment.y -= SCALE_STEP;
-    } else if key_input.just_pressed(KeyCode::Key8) {
-        projection_adjustment.y += SCALE_STEP;
-    } else if key_input.just_pressed(KeyCode::Key9) {
-        projection_adjustment.z -= SCALE_STEP;
-    } else if key_input.just_pressed(KeyCode::Key0) {
-        projection_adjustment.z += SCALE_STEP;
-    }
```
The instructions also need to be removed from the `impl fmt::Display for SceneHandle`.

danchia replied:
Done.

superdump reviewed (changes requested):
Initial pass.

superdump commented:
Performance seems to be very poor on this PR with shadows enabled even on a fairly simple scene like the flight helmet using the `scene_viewer` example without any arguments.

danchia commented:
> 

Interesting, I will take a closer look as to why

danchia commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -1168,47 +1199,73 @@ pub fn prepare_lights(
         // directional lights
+        let mut directional_depth_texture_array_index = 0u32;
         for (light_index, &(light_entity, light)) in directional_lights
             .iter()
             .enumerate()
-            .take(directional_shadow_maps_count)
+            .take(directional_shadow_enabled_count)
```
There is a sort above that sorts shadow-enabled lights first

superdump commented:
As a note, this is on an M1 Max. I'm continuing investigations on Discord as it will be noisy to spam random findings here. :)

superdump commented:
With changes like those in #7069 also for the scope of this PR we were able to keep performance in check. :)

danchia commented on crates/bevy_pbr/src/light.rs:
```diff
@@ -256,9 +227,243 @@ pub struct DirectionalLightShadowMap {
+    let views = views
+        .iter()
+        .filter_map(|x| match x {
+            // TODO: orthographic camera projection support.
```
Agreed that it will be as large close as far, but it may still make sense to devote more pixels to closer distances than further ones w.r.t to quality bang for buck :)

danchia commented on crates/bevy_pbr/src/render/mesh_view_types.wgsl:
```diff
@@ -28,14 +28,23 @@ struct PointLight {
-struct DirectionalLight {
+struct DirectionalCascade {
     view_projection: mat4x4<f32>,
+    texel_size: vec4<f32>,
```
Done

danchia commented:
I made changes to the shader to follow what https://github.com/bevyengine/bevy/pull/7069 did specifically for directional shadowing code that I modified, left the rest of the optimizations to that PR.

superdump commented:
I'll give this another review pass when I can.

mockersf commented:
I tried comparing example `lighting` for main and this pr

https://user-images.githubusercontent.com/8672791/210342283-8e1aaa7b-90a3-45d7-9a7a-620d470915ac.mp4

Shadow borders are sometime less clean, and there's something strange happening in the back at 5 seconds

superdump commented:
Good catch @mockersf ! The aliased edges is a cascade bound tweaking issue. By default we're configuring 4 cascades with the first cascade far bound at 5 units, and the last at 1000 units. This scene is small so 1000 units is way too much. But also we probably just want to adjust a bit to something that looks good out of the box with the current 2x2 Percentage Closer Filtering (using multiple samples of the shadow map to filter the edges).

However, the bug in the back behind the wall is something we will need to fix. It looks like the red and blue light are making it through the wall...
<img width="574" alt="Screenshot 2023-01-03 at 12 05 01" src="https://user-images.githubusercontent.com/302146/210345368-ffbb4434-49e1-45d2-9fd1-4d9385e1826a.png">
The red and blue lights are point lights.

mockersf commented:
WASM also has something funny:

https://user-images.githubusercontent.com/8672791/210359269-48057ce0-d50e-44b3-9e4b-0dfd257069b0.mp4

danchia commented:
> WASM also has something funny:
> 
>  cascaded-wasm.mp4

Ah, for WASM (which I assume uses the `webgl` feature) we should probably change the default cascade config to only 1 cascade, and some more reasonable far bound that 1000. webgl doesn't have texture arrays, so we can only use one cascade.

But I think WASM users are going to have to manually configure the far bound for best results still. since achieving a tight bound is important for quality vs shadow map resolution.

danchia commented:
I will look into what's going on with the point lights bleeding through the wall.

danchia commented:
* Fixed the weird shadow light bleeding problems. Turns out my depth clamp (as written) only works on ortho matrices, so I changed the code to only turn on the clamping for directional lights.
* Updating lighting example to show how / when to customize the cascades config
* Updated the default cascades config for webgl to only use 1 cascade, covering a smaller shadow distance.


https://user-images.githubusercontent.com/1222141/211190615-4a1faf5d-f885-4840-a799-4dac6a67a3d6.mp4

superdump commented on crates/bevy_pbr/src/render/shadows.wgsl:
```diff
@@ -127,8 +139,42 @@ fn fetch_directional_shadow(light_id: u32, frag_position: vec4<f32>, surface_nor
+
+    var shadow = sample_cascade(light_id, cascade_index, frag_position, surface_normal);
+
+    // Blend with the next cascade, if there's one.
```
```suggestion
    // Blend with the next cascade, if there is one.
```

superdump commented on examples/3d/load_gltf.rs:
```diff
@@ -21,18 +21,8 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
     });
-    const HALF_SIZE: f32 = 1.0;
     commands.spawn(DirectionalLightBundle {
         directional_light: DirectionalLight {
```
Probably also tweak the cascades here so this one looks good. It's kind of the fanciest-looking in-repo PBR example.

superdump reviewed (commented):
This looks pretty good to me. I'm going to do a bit more testing locally but my finger is hovering over approval. We can rebase after #6284 is merged.

rparrett commented:
Should there be a warning or something when `CascadeShadowConfig` has `num_cascades` > `MAX_CASCADES_PER_LIGHT`?

In a test scene where the camera was pretty far from the action, bumping value up that resulted in shadows mysteriously disappearing (as only the first 4 cascades were used?).

danchia commented:
> Should there be a warning or something when `CascadeShadowConfig` has `num_cascades` > `MAX_CASCADES_PER_LIGHT`?
> 
> In a test scene where the camera was pretty far from the action, bumping value up that resulted in shadows mysteriously disappearing (as only the first 4 cascades were used?).

That's a good idea, will look into it.

robtfm commented on crates/bevy_pbr/src/light.rs:
```diff
@@ -256,9 +227,249 @@ pub struct DirectionalLightShadowMap {
+        shadow_maximum_distance: f32,
+        overlap_proportion: f32,
+    ) -> Self {
+        assert!(num_cascades > 0, "num_cascades must be positive");
```
please can we assert `overlap_proportion < 1` as well? >1 causes it to hit the z_near assert below but is a little more cryptic.

mockersf replied:
~~`num_cascades` is an size, it can't be negative~~
oh that's actually a cultural thing, in French 0 is positive but not in English

robtfm commented on examples/tools/scene_viewer/scene_viewer_plugin.rs:
```diff
@@ -44,9 +44,6 @@ Scene Controls:
     L           - animate light direction
     U           - toggle shadows
     C           - cycle through the camera controller and any cameras loaded from the scene
-    5/6         - decrease/increase shadow projection width
```
could we keep some inputs for updating the cascade far and near settings?

robtfm replied:
this is not blocking, could be a follow up

superdump replied:
Good idea. We could either allow control of just the number of cascades and the first and last far bounds and have the code calculate the bounds of each cascade, or allow full control. As it is meant to be illustrative, I'm inclined to do the former as it's a bit simpler until we have some UI slider stuff.

danchia replied:
sounds good, I will work on this as a follow up if that's ok with you

robtfm commented on crates/bevy_render/src/view/mod.rs:
```diff
@@ -87,6 +87,7 @@ impl Default for Msaa {
 pub struct ExtractedView {
     pub projection: Mat4,
     pub transform: GlobalTransform,
+    pub view_projection: Option<Mat4>,
```
would it be possible to remove the old projection field and just do the projection * inverse calculation using `view_projection` wherever `projection` is currently used? then you have the fine control you need without duplicating the data?

if not then can we add some comments here. it seems like a source of potential future confusion to have a `projection` and a `view_projection` field with no descriptions.

robtfm replied:
after discussing with robswain on discord, it's fine to keep both, pls disregard this.

danchia replied:
Keeping field, but I added a comment explaining why it's optional and when you'd want to use it.

danchia commented on crates/bevy_pbr/src/light.rs:
```diff
@@ -1462,34 +1674,38 @@ fn project_to_plane_y(y_light: Sphere, y_plane: Plane, is_orthographic: bool) ->
+        frusta.frusta = cascades
+            .cascades
+            .iter()
+            .map(|(view, cascades)| {
+                (
+                    *view,
+                    cascades
+                        .iter()
+                        .map(|c| Frustum::from_view_projection(&c.view_projection))
+                        .collect::<Vec<_>>(),
+                )
+            })
```
nope, cargo fmt puts it back that way.

danchia commented on crates/bevy_pbr/src/render/depth.wgsl:
```diff
@@ -38,5 +38,9 @@ fn vertex(vertex: Vertex) -> VertexOutput {
 
     var out: VertexOutput;
     out.clip_position = mesh_position_local_to_clip(model, vec4<f32>(vertex.position, 1.0));
+    #ifdef DEPTH_CLAMP_ORTHO
+        out.clip_position.z = min(out.clip_position.z, 1.0);
+    #endif
```
ah thanks for showing me the convention, fixed.

danchia commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -1047,7 +1079,8 @@ pub fn prepare_lights(
             // the spot lights themselves start in the light array at point_light_count. so to go from light
             // index to shadow map index, we need to subtract point light count and add directional shadowmap count.
-            spot_light_shadowmap_offset: directional_shadow_maps_count as i32
+            spot_light_shadowmap_offset: (directional_shadow_enabled_count * MAX_CASCADES_PER_LIGHT)
```
Good catch, changed to use `num_directional_cascades_enabled` to mirror how we set it on line 1173 where we assign the `depth_texture_view` for spot light shadow map rendering.

danchia commented:
I believe feedback is addressed, with one exception (noted below). Remaining items:
* tweak `load_gltf` to look nice.
* rebase against latest master

danchia commented:
I've tweaked load_gltf to shrink the distances which improves quality, and have rebased.

rparrett commented:
I have a couple doc suggestions.

In a scenario like a 2.5d/3d strategy game where the camera is potentially very far away from the action, is my only option for increasing shadow map waste tweaking `nearest_bound`?

And am I right that this is effectively only pushing the end of the first cascade, potentially covering my objects with more cascades, but still wasting the entire first cascade?

Maybe this diagram will help:

```
Camera | Cascade 1          | C2       | C3  | C4 | shadow_maximum_distance
Camera | Cascade 1                     |C2 |C3 |C4| shadow_maximum_distance
                                        <<Things>>
```

1. As a user, it feels a bit unintuitive. The docs aren't incorrect or anything, but i feel like spelling out the fact that `nearest_bound` is not `minimum_shadow_distance` more explicitly might have helped me.
2. On WASM, or when there's otherwise only 1 cascade, `nearest_bound` is ignored and this doesn't seem to be mentioned anywhere.

rparrett commented:
Also, is there some reason I'm missing for why we can't and/or don't really need to configure a `shadow_minimum_distance`?

rparrett commented:
That makes sense, thanks. I do think the docs could be a bit clearer on this.

superdump commented:
> In a scenario like a 2.5d/3d strategy game where the camera is potentially very far away from the action, is my only option for increasing shadow map waste tweaking `nearest_bound`?

`CascadeShadowConfig`'s members are public. You can construct the far bounds of the cascades however you like. It's just a `Vec<f32>`. However, the `new()` aims to be a useful convenience constructor given the usual concern of trading off performance and quality via the number of cascades and the shadow far distance. Setting the far bound of the nearest cascade is to avoid spending a bunch of cascades on very small distance ranges close to the camera.

I hear you that for a top-down/isometric type game, a different approach might indeed be useful. However, I think given there is low distribution of objects over camera depth versus distribution in screen space, the only thing that will really matter for such cases would be shadow map resolution. Then you could use a single cascade and higher shadow map texture resolution.

> And am I right that this is effectively only pushing the end of the first cascade, potentially covering my objects with more cascades, but still wasting the entire first cascade?
> 
> Maybe this diagram will help:
> 
> ```
> Camera | C1 | C2   | C3      | C4        | shadow_maximum_distance
> Camera | C1          | C2 | C3  | C4     | shadow_maximum_distance (nearest_bound tweaked)
>                      <<Things           >>
> ```

That is what adjusting the nearest bound would do, yes.

>     1. As a user, it feels a bit unintuitive. The docs aren't incorrect or anything, but i feel like spelling out the fact that `nearest_bound` is not `minimum_shadow_distance` more explicitly might have helped me.

Good point. And perhaps we could elaborate with suggestions of when to use multiple cascades and exceptions of when to use fewer or one or so.

>     2. On WASM, or when there's otherwise only 1 cascade, `nearest_bound` is ignored and this doesn't seem to be mentioned anywhere.

Also a good point, thanks for the feedback!

superdump commented:
I suppose with an isometric-type view with the camera high above the action, the closest object to the camera would be at a greater distance from the camera and then as you pointed out, it would be good to not try to encompass the camera origin in the shadow mapping. Which then goes to what robtfm was saying about moving the projection's near plane, which defines the view frustum. It is probably worth someone making a test case for it and trying to put together something that works well. It's not completely clear what the recommendation would be though and I don't really want to hold up this PR. We can iterate on documentation I think.

superdump reviewed (approved):
@robtfm if you're ok with this then I think we can merge it.

robtfm reviewed (approved):
@superdump r+

superdump commented:
I’ll open an issue about improving the documentation for different cases.

superdump commented:
bors r+
